pkg/rehearse/jobs.go

```diff
@@ -222,6 +228,107 @@ func ConfigureRehearsalJobs(toBeRehearsed config.Presubmits, ciopConfigs config.
 	return rehearsals
 }
 
+// AddRandomJobsForChangedTemplates finds random jobs from the PR config that are using a specific template with a specific cluster type.
+// So if a template will be changed, find the jobs that are using a template in combination with the `aws`,`openstack`,`gcs` and `libvirt` cluster types.
+func AddRandomJobsForChangedTemplates(templatesWithClusterTypes map[string]sets.String, templates config.CiTemplates, prConfigPresubmits map[string][]prowconfig.Presubmit, loggers Loggers, prNumber int) config.Presubmits {
+	rehearsals := make(config.Presubmits)
+
+	for templateFile, clusterTypes := range templatesWithClusterTypes {
+		for clusterType := range clusterTypes {
+			if repo, job := pickTemplateJob(prConfigPresubmits, templateFile, clusterType); job != nil {
+				jobLogger := loggers.Job.WithFields(logrus.Fields{"target-repo": repo, "target-job": job.Name})
+
+				volumeRef, err := getVolumeIndexAndTempCMName(job.Spec.Containers[0].VolumeMounts, job.Spec.Volumes, templates)
+				if err != nil {
+					jobLogger.WithError(err).Warn(skipJobMessage)
+					continue
+				}
+
+				if volumeRef != nil {
+					job.Spec.Volumes[volumeRef.index].VolumeSource.ConfigMap.Name = volumeRef.name
+				}
+
+				rehearsals[repo] = append(rehearsals[repo], *job)
+			}
+		}
+	}
+	return rehearsals
+}
+
+type volumeRef struct {
+	index       int
+	name        string
+	templateKey string
+}
+
+func getVolumeIndexAndTempCMName(volumeMounts []v1.VolumeMount, volumes []v1.Volume, templates config.CiTemplates) (*volumeRef, error) {
+	exists, index, templateKey := hasChangedTemplateVolume(volumeMounts, volumes, templates)
+	if !exists {
+		return nil, nil
+	}
+
+	templateData, err := config.GetTemplateData(templates[templateKey])
```

Comment:
We decode the same templates (in `config.GetTemplateData`) countless times. We should store the value when they are first loaded. Many of the new "error handling" `if`s we have are because of that function, even though it is guaranteed not to fail if it succeeded once.

Reply:
that's sound like a good follow-up